test/test_mcmc.py

```diff
@@ -127,7 +127,10 @@ def test_improper_normal():
 
     def model(data):
         alpha = numpyro.sample('alpha', dist.Uniform(0, 1))
-        loc = numpyro.param('loc', 0., constraint=constraints.interval(0., alpha))
+        with reparam(config={'loc': TransformReparam()}):
```

Comment:
We need the `reparam` for the dynamic upper bound `alpha`, right?

Purely from an interface point of view, this seems like a regression. :) This is probably not a pattern that we would like to advertise. I still appreciate that we can clear off a lot of our tech debt, and if anyone wants to have optimizable params with dynamic support, there is a way to do this. This seems like a fringe enough use case that it is better to opt for code simplicity.

Reply:
You are right, this is a regression. However, the change simplifies much of code, it makes things less confusing (there are several issues come from the confusing base_param_map, substitute_fn only applies for base value,...). We also gain some performance with the change.

About this improper prior, there are two ways of thinking:
+ flat prior over (0, 1)
+ flat prior over (0, alpha)
Before this change, the behavior is flat prior over (0, 1). Is it the behavior that we want? I am not sure. With this change, it is clear that users want a flat prior over (0, 1). To get a flat prior over (0, alpha), I think we just need to add `handlers.mask(False)` context out of this reparam context. So things are more explicit to me. :D

Reply:
I read this as flat prior over (0, alpha). I would imagine that something like the following should work. Is that not the case?

```python
with reparam(config={'loc': TransformReparam()}):
    loc = numpyro.sample('loc', dist.Uniform(0, alpha).mask(False))
```

Reply:
The issue is TransformReparam does not work with `MaskedDistribution` yet. So you would need to use `mask` handler here. I'll make an issue for it. After that, your code will work as you expected.